lib/storage/storage.go

```diff
@@ -244,6 +249,22 @@ func MustOpenStorage(path string, retention time.Duration, maxHourlySeries, maxD
 	isEmptyDB := !fs.IsPathExist(filepath.Join(path, indexdbDirname))
 	fs.MustMkdirIfNotExist(metadataDir)
 	s.minTimestampForCompositeIndex = mustGetMinTimestampForCompositeIndex(metadataDir, isEmptyDB)
+	nodeIDPath := filepath.Join(metadataDir, nodeIDFilename)
+	if fs.IsPathExist(nodeIDPath) {
+		r, err := os.Open(nodeIDPath)
+		if err != nil {
+			logger.Panicf("FATAL: cannot open nodeID file %q: %s", nodeIDPath, err)
+		}
+
+		nodeID, err := io.ReadAll(r)
+		if err != nil {
+			logger.Panicf("FATAL: cannot read nodeID from %q: %s", nodeIDPath, err)
+		}
+		s.nodeID = encoding.UnmarshalUint64(nodeID)
+	} else {
+		nodeID := rand.Uint64()
+		s.nodeID = nodeID
```

Comment:
Looks like new id must be saved immedeatly after it was generated with s.mustSaveNodeID()

Otherwise, it could be lost due to server crash.

 What is a chance of collision for math/rand?

Reply:
how to reset id if needed?

Reply:
> What is a chance of collision for math/rand? 

`math/rand` is pseudo-random which is initualized with random source since Go 1.20 release, so we shouldn't expect any collisions here.

> how to reset id if needed?

Just delete `metadata/node_id.bin` and restart the storage. That would trigger generation of node ID again.

Reply:
In this case, `mustSaveNodeID` cannot be called at storage shutdown (at first init is enough). Otherwise it'll override value and ll make it impossible to reset id for kubernetes or similar envs.

Reply:
While it does have lower chance of collision I would rather think if we really need the cance to be lower than uint64 already provides?
With 100 vmstorage nodes chance of collision will be 2.220446049250313e-16 (based on approximation of [Birthday problem](https://en.wikipedia.org/wiki/Birthday_problem)). It seems to me like having UUID here (122 random bits for v4 UUID) would not be that critical considering that node ID is generated once and cluster size is rarely exceeding even hundreds of nodes.